File: labs/M13-planning-task-decomposition/starter/planning_agent.py
"""
M13 Lab — Planning & Task Decomposition (Starter)
===================================================
Build a planning agent that decomposes complex queries into
ordered sub-tasks, then executes each step using a ReAct loop.

KEY CONCEPT: The "plan-then-act" pattern separates THINKING
from DOING. First, ask Claude to create a plan (list of steps
with dependencies). Then execute each step one at a time,
feeding results from earlier steps into later ones. This is
more reliable than letting the agent improvise on every turn.

Usage:
    python planning_agent.py
"""

import json
import sys
import os
from dotenv import load_dotenv

load_dotenv()

sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))
sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", ".."))

import anthropic
from shared.mock_ucc_data import search_filings, get_filing_by_number

client = anthropic.Anthropic()
MODEL = "claude-sonnet-4-6"


# =============================================================================
# OBSERVATION HELPERS (complete — do not modify)
# =============================================================================

def observe(label: str, message: str) -> None:
    print(f"\n{'=' * 60}")
    print(f"[{label}] {message}")
    print(f"{'=' * 60}")


def observe_plan(plan: list[dict]) -> None:
    print(f"\n{'=' * 60}")
    print("[PLAN] Decomposed into steps:")
    for i, step in enumerate(plan, 1):
        deps = step.get("depends_on", [])
        dep_str = f" (depends on: {', '.join(deps)})" if deps else ""
        print(f"  {i}. {step['task']}{dep_str}")
    print(f"{'=' * 60}")


def observe_step(step_num: int, task: str) -> None:
    print(f"\n{'─' * 60}")
    print(f"[STEP {step_num}] Executing: {task}")
    print(f"{'─' * 60}")


def observe_step_result(step_num: int, result: str) -> None:
    print(f"\n{'─' * 60}")
    print(f"[STEP {step_num} RESULT]")
    if len(result) > 300:
        print(result[:300] + "\n... (truncated)")
    else:
        print(result)
    print(f"{'─' * 60}")


# =============================================================================
# TOOL DEFINITIONS (complete — do not modify)
# =============================================================================

TOOLS = [
    {
        "name": "search_filings",
        "description": "Search UCC filings by debtor name and/or state.",
        "input_schema": {
            "type": "object",
            "properties": {
                "debtor_name": {"type": "string", "description": "Debtor name to search for"},
                "state": {"type": "string", "description": "State to filter by"}
            },
            "required": []
        }
    },
    {
        "name": "get_filing_details",
        "description": "Get full details of a specific UCC filing by filing number.",
        "input_schema": {
            "type": "object",
            "properties": {
                "filing_number": {"type": "string", "description": "The UCC filing number"}
            },
            "required": ["filing_number"]
        }
    },
    {
        "name": "calculate_risk",
        "description": "Calculate risk profile for a debtor based on their UCC filings.",
        "input_schema": {
            "type": "object",
            "properties": {
                "debtor_name": {"type": "string", "description": "Debtor name to assess"}
            },
            "required": ["debtor_name"]
        }
    },
    {
        "name": "generate_report_section",
        "description": "Generate a section of a risk report. Provide a section title and the data to summarize.",
        "input_schema": {
            "type": "object",
            "properties": {
                "section_title": {"type": "string", "description": "Title for this report section"},
                "data": {"type": "string", "description": "Raw data/findings to summarize in this section"}
            },
            "required": ["section_title", "data"]
        }
    }
]


# =============================================================================
# TOOL EXECUTION (complete — do not modify)
# =============================================================================

def calculate_risk_for_debtor(debtor_name: str) -> dict:
    """Analyze all filings for a debtor and return a risk profile."""
    filings = search_filings(debtor_name=debtor_name)
    if not filings:
        return {"debtor": debtor_name, "risk_score": 0, "risk_level": "UNKNOWN",
                "message": f"No filings found for '{debtor_name}'"}
    active = [f for f in filings if f["status"] == "Active"]
    blanket = [f for f in filings if "all assets" in f["collateral_description"].lower()
               or "all accounts" in f["collateral_description"].lower()]
    amendments = [f for f in filings if f["type"] == "UCC-3"]
    score = min(1.0, (len(active) * 0.25) + (len(blanket) * 0.3) + (len(amendments) * 0.1))
    if score >= 0.7:
        level, rec = "HIGH", "Significant lien exposure. Detailed due diligence required."
    elif score >= 0.4:
        level, rec = "MEDIUM", "Moderate lien activity. Review collateral descriptions."
    else:
        level, rec = "LOW", "Limited lien exposure. Standard procedures sufficient."
    return {"debtor": debtor_name, "risk_score": round(score, 2), "risk_level": level,
            "total_filings": len(filings), "active_filings": len(active),
            "blanket_liens": len(blanket), "amendments": len(amendments),
            "recommendation": rec}


def execute_tool(tool_name: str, tool_input: dict) -> str:
    """Dispatch a tool call. Returns JSON string."""
    try:
        if tool_name == "search_filings":
            results = search_filings(
                debtor_name=tool_input.get("debtor_name"),
                state=tool_input.get("state")
            )
            return json.dumps([{
                "filing_number": f["filing_number"],
                "debtor": f["debtor"]["name"],
                "secured_party": f["secured_party"]["name"],
                "state": f["state"], "status": f["status"], "type": f["type"],
                "collateral": f["collateral_description"][:120] + "..."
            } for f in results], indent=2)
        elif tool_name == "get_filing_details":
            filing = get_filing_by_number(tool_input["filing_number"])
            if filing:
                return json.dumps(filing, indent=2, default=str)
            return json.dumps({"error": f"Filing {tool_input['filing_number']} not found"})
        elif tool_name == "calculate_risk":
            return json.dumps(calculate_risk_for_debtor(tool_input["debtor_name"]), indent=2)
        elif tool_name == "generate_report_section":
            # This tool just passes the data through — Claude synthesizes in context
            return json.dumps({
                "section": tool_input["section_title"],
                "content": f"[Report section '{tool_input['section_title']}' generated from provided data]",
                "data_received": tool_input["data"][:200] + "..."
            }, indent=2)
        else:
            return json.dumps({"error": f"Unknown tool: {tool_name}"})
    except Exception as e:
        return json.dumps({"error": f"Tool failed: {str(e)}"})


# =============================================================================
# PLANNING AGENT — YOUR CODE HERE
# =============================================================================

def create_plan(user_query: str) -> list[dict]:
    """
    Ask Claude to decompose a complex query into ordered sub-tasks.

    Returns a list of step dicts, each with:
    - "step_id": unique identifier (e.g. "step_1")
    - "task": description of what this step does
    - "depends_on": list of step_ids that must complete first (can be empty)
    - "tools_needed": list of tool names this step will likely use

    Example output:
    [
        {"step_id": "step_1", "task": "Search for all UCC filings for the company",
         "depends_on": [], "tools_needed": ["search_filings"]},
        {"step_id": "step_2", "task": "Get full details for each filing found",
         "depends_on": ["step_1"], "tools_needed": ["get_filing_details"]},
        ...
    ]
    """
    # ------------------------------------------------------------------
    # TODO 1: Implement create_plan()
    #   - Call client.messages.create() with a system prompt that instructs
    #     Claude to decompose the query into 3-6 ordered steps
    #   - The system prompt should specify the output format (JSON array)
    #   - Parse the JSON from Claude's response
    #   - Return the list of step dicts
    #   - If parsing fails, return a single-step fallback plan
    #
    # Hint for the system prompt:
    #   "You are a task planning agent. Given a complex query, decompose it
    #    into 3-6 ordered steps. Return ONLY a JSON array of step objects.
    #    Each step has: step_id, task, depends_on (list of step_ids), tools_needed.
    #    Available tools: search_filings, get_filing_details, calculate_risk,
    #    generate_report_section."
    # ------------------------------------------------------------------
    pass


def execute_step(step: dict, context: str, max_turns: int = 5) -> str:
    """
    Execute a single plan step using the ReAct loop from M12.

    Args:
        step: The step dict from the plan
        context: Results from previous steps (passed as additional context)
        max_turns: Max tool-use iterations for this step

    Returns:
        The text result from executing this step
    """
    # ------------------------------------------------------------------
    # TODO 2: Implement execute_step()
    #   - Build a system prompt that includes:
    #     a) The step's task description
    #     b) Context from previous steps
    #     c) Instructions to use tools as needed and return a concise result
    #   - Run a ReAct loop (like M12) for max_turns iterations
    #   - Return the final text response
    # ------------------------------------------------------------------
    pass


def run_planning_agent(user_query: str) -> str:
    """
    Run the full planning agent:
    1. Create a plan (decompose query into steps)
    2. Execute each step in order, passing results forward
    3. Return the final synthesized result

    Args:
        user_query: The user's complex query

    Returns:
        The final combined result from all steps
    """
    observe("QUERY", user_query)

    # ------------------------------------------------------------------
    # TODO 3: Implement run_planning_agent()
    #   - Call create_plan(user_query) to get the step list
    #   - Log the plan with observe_plan(plan)
    #   - Initialize a results dict and context string
    #   - For each step in the plan:
    #     a) Log with observe_step(step_num, step["task"])
    #     b) Call execute_step(step, context)
    #     c) Log with observe_step_result(step_num, result)
    #     d) Store the result and append to context
    #   - After all steps, synthesize a final answer
    #     (either use the last step's result or call Claude once more)
    #   - Return the final answer
    # ------------------------------------------------------------------
    pass


# =============================================================================
# MAIN
# =============================================================================

if __name__ == "__main__":
    print("=" * 60)
    print("M13 Lab — Planning & Task Decomposition")
    print("=" * 60)

    # Scenario 1: Single-entity risk report
    print("\n\n>>> Scenario 1: Generate risk report")
    result1 = run_planning_agent(
        "Generate a complete risk report for Greenfield Logistics LLC"
    )
    print(f"\nFINAL REPORT:\n{result1}")

    # Scenario 2: Comparative analysis (requires parallel-ish research)
    print("\n\n>>> Scenario 2: Compare two entities")
    result2 = run_planning_agent(
        "Compare lien exposure between Nextera Holdings and Lone Star Energy"
    )
    print(f"\nFINAL REPORT:\n{result2}")

File: labs/M13-planning-task-decomposition/starter/planning_agent.js
/**
 * M13 Lab — Planning & Task Decomposition (Starter)
 * ===================================================
 * Build a planning agent that decomposes complex queries into
 * ordered sub-tasks, then executes each step using a ReAct loop.
 *
 * Usage:
 *     node planning_agent.js
 */

import Anthropic from "@anthropic-ai/sdk";
import { config } from "dotenv";
import { dirname, join } from "path";
import { fileURLToPath, pathToFileURL } from "url";

config();

const __dirname = dirname(fileURLToPath(import.meta.url));
// pathToFileURL: ESM import() needs a file:// URL, not a filesystem path.
// A bare Windows path (D:\...) is read as protocol 'd:' and rejected with
// ERR_UNSUPPORTED_ESM_URL_SCHEME. Posix paths happen to work, which is why
// this only breaks on Windows.
const { searchFilings, getFilingByNumber } = await import(pathToFileURL(join(__dirname, "..", "..", "shared", "mock_ucc_data.js")).href);

const client = new Anthropic();
const MODEL = "claude-sonnet-4-6";

// =============================================================================
// OBSERVATION HELPERS (complete — do not modify)
// =============================================================================

function observe(label, message) {
  console.log(`\n${"=".repeat(60)}`);
  console.log(`[${label}] ${message}`);
  console.log("=".repeat(60));
}

function observePlan(plan) {
  console.log(`\n${"=".repeat(60)}`);
  console.log("[PLAN] Decomposed into steps:");
  plan.forEach((step, i) => {
    const deps = step.depends_on || [];
    const depStr = deps.length ? ` (depends on: ${deps.join(", ")})` : "";
    console.log(`  ${i + 1}. ${step.task}${depStr}`);
  });
  console.log("=".repeat(60));
}

function observeStep(num, task) {
  console.log(`\n${"─".repeat(60)}`);
  console.log(`[STEP ${num}] Executing: ${task}`);
  console.log("─".repeat(60));
}

function observeStepResult(num, result) {
  console.log(`\n${"─".repeat(60)}`);
  console.log(`[STEP ${num} RESULT]`);
  console.log(result.length > 300 ? result.slice(0, 300) + "\n... (truncated)" : result);
  console.log("─".repeat(60));
}

// =============================================================================
// TOOLS (complete — do not modify)
// =============================================================================

const TOOLS = [
  { name: "search_filings", description: "Search UCC filings by debtor name and/or state.",
    input_schema: { type: "object", properties: { debtor_name: { type: "string" }, state: { type: "string" } }, required: [] } },
  { name: "get_filing_details", description: "Get full details of a specific UCC filing.",
    input_schema: { type: "object", properties: { filing_number: { type: "string" } }, required: ["filing_number"] } },
  { name: "calculate_risk", description: "Calculate risk profile for a debtor.",
    input_schema: { type: "object", properties: { debtor_name: { type: "string" } }, required: ["debtor_name"] } },
  { name: "generate_report_section", description: "Generate a report section from data.",
    input_schema: { type: "object", properties: { section_title: { type: "string" }, data: { type: "string" } }, required: ["section_title", "data"] } },
];

function calculateRiskForDebtor(name) {
  const filings = searchFilings({ debtorName: name });
  if (!filings.length) return { debtor: name, risk_score: 0, risk_level: "UNKNOWN" };
  const active = filings.filter(f => f.status === "Active");
  const blanket = filings.filter(f => f.collateral_description.toLowerCase().includes("all assets") || f.collateral_description.toLowerCase().includes("all accounts"));
  const amend = filings.filter(f => f.type === "UCC-3");
  const score = Math.min(1.0, active.length * 0.25 + blanket.length * 0.3 + amend.length * 0.1);
  const level = score >= 0.7 ? "HIGH" : score >= 0.4 ? "MEDIUM" : "LOW";
  return { debtor: name, risk_score: Math.round(score * 100) / 100, risk_level: level,
    total_filings: filings.length, active_filings: active.length, blanket_liens: blanket.length, amendments: amend.length };
}

function executeTool(name, input) {
  try {
    if (name === "search_filings") {
      const r = searchFilings({ debtorName: input.debtor_name, state: input.state });
      return JSON.stringify(r.map(f => ({ filing_number: f.filing_number, debtor: f.debtor.name, secured_party: f.secured_party.name, state: f.state, status: f.status, type: f.type, collateral: f.collateral_description.slice(0,120)+"..." })), null, 2);
    } else if (name === "get_filing_details") {
      const f = getFilingByNumber(input.filing_number);
      return f ? JSON.stringify(f, null, 2) : JSON.stringify({ error: "Not found" });
    } else if (name === "calculate_risk") {
      return JSON.stringify(calculateRiskForDebtor(input.debtor_name), null, 2);
    } else if (name === "generate_report_section") {
      return JSON.stringify({ section: input.section_title, content: `[Section generated]`, data_received: (input.data||"").slice(0,200)+"..." }, null, 2);
    }
    return JSON.stringify({ error: `Unknown tool: ${name}` });
  } catch (e) { return JSON.stringify({ error: e.message }); }
}

// =============================================================================
// PLANNING AGENT — YOUR CODE HERE
// =============================================================================

/**
 * Ask Claude to decompose a complex query into ordered sub-tasks.
 *
 * TODO 1: Implement this function
 *   - Call client.messages.create() with a planning system prompt
 *   - Parse the JSON plan from Claude's response
 *   - Return array of step objects: [{step_id, task, depends_on, tools_needed}]
 */
async function createPlan(userQuery) {
  // Your code here
  return [{ step_id: "step_1", task: userQuery, depends_on: [], tools_needed: ["search_filings"] }];
}

/**
 * Execute a single plan step using the ReAct loop.
 *
 * TODO 2: Implement this function
 *   - Build a system prompt with the step task and previous context
 *   - Run a ReAct loop (tool_use check → execute → loop)
 *   - Return the final text result
 */
async function executeStep(step, context, maxTurns = 5) {
  // Your code here
  return "TODO: Implement step execution";
}

/**
 * Run the full planning agent: plan → execute → synthesize.
 *
 * TODO 3: Implement this function
 *   - createPlan() to decompose the query
 *   - Loop through steps, calling executeStep() for each
 *   - Build cumulative context
 *   - Synthesize final result
 */
async function runPlanningAgent(userQuery) {
  observe("QUERY", userQuery);
  // Your code here
  return "TODO: Implement planning agent";
}

// =============================================================================
// MAIN
// =============================================================================

console.log("=".repeat(60));
console.log("M13 Lab — Planning & Task Decomposition");
console.log("=".repeat(60));

console.log("\n\n>>> Scenario 1: Generate risk report");
const r1 = await runPlanningAgent("Generate a complete risk report for Greenfield Logistics LLC");
console.log(`\nFINAL REPORT:\n${r1}`);

console.log("\n\n>>> Scenario 2: Compare two entities");
const r2 = await runPlanningAgent("Compare lien exposure between Nextera Holdings and Lone Star Energy");
console.log(`\nFINAL REPORT:\n${r2}`);
